subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskTimeoutIntegrationTest.groovy

```diff
@@ -288,4 +294,86 @@ class TaskTimeoutIntegrationTest extends AbstractIntegrationSpec {
         def timeoutLogging = logging.findAll { it.category == DefaultTimeoutHandler.name }
         timeoutLogging.collect { it.message } as List<String>
     }
+
+    def "task is re-interrupted until it stops"() {
+        given:
+        postTimeoutCheckFrequencyMs = 100
+        buildFile << """
+            task block() {
+                doLast {
+                    def interruptedCount = 0
+                    while (true) {
+                        if (Thread.interrupted()) {
+                            println "received interrupt"
+                            if (++interruptedCount == 3) {
+                                return
+                            }
+                        }
+                    }
+                }
+                timeout = Duration.ofMillis($TIMEOUT)
+            }
+            """
+
+        expect:
+        fails "block"
+
+        and:
+        with(result.groupedOutput.task(":block")) {
+            output.count("received interrupt") == 3
+        }
+    }
+
+    def "stack trace of task is printed if it is slow to stop"() {
+        given:
+        executer.withStackTraceChecksDisabled()
+        postTimeoutCheckFrequencyMs = 100
+        slowStopLogStacktraceFrequencyMs = 100
+
+        buildFile << """
+            @groovy.transform.CompileStatic
```

Comment:
Why the `CompileStatic`? It makes the test harder to read.

Reply:
It reduces the volatility in the traces, making the test output/results a bit easier to reason about.

Reply:
👍 

Does that maybe mean we should filter the stacktraces we print out for the threads so we remove the groovy dynamic invocation parts?

Reply:
I tried that, but it's a bit involved. This module can't see the filterer.